File: tests/test_edge_geometry.cpp
#include <catch2/catch_test_macros.hpp>
#include "gravel/core/edge_geometry.h"
#include "gravel/simplify/simplify.h"

#include <vector>

using namespace gravel;

// Bidirectional path 0-1-2-3-4 with distinct coordinates. Nodes 1,2,3 are
// degree-2 (contractible); 0 and 4 are the surviving junctions.
static ArrayGraph make_path5_with_coords() {
    std::vector<uint32_t> offsets = {0, 1, 3, 5, 7, 8};
    std::vector<NodeID> targets = {1, 0, 2, 1, 3, 2, 4, 3};
    std::vector<Weight> weights(targets.size(), 1.0);
    std::vector<Coord> coords = {
        {0.0, 0.0}, {0.0, 1.0}, {0.0, 2.0}, {0.0, 3.0}, {0.0, 4.0}};
    return ArrayGraph(std::move(offsets), std::move(targets),
                      std::move(weights), std::move(coords));
}

// Star: center 0 (degree 3) with leaves 1,2,3 (degree 1). Nothing contracts,
// so every edge is a direct junction-to-junction segment.
static ArrayGraph make_star_with_coords() {
    std::vector<uint32_t> offsets = {0, 3, 4, 5, 6};
    std::vector<NodeID> targets = {1, 2, 3, 0, 0, 0};
    std::vector<Weight> weights(targets.size(), 1.0);
    std::vector<Coord> coords = {
        {0.0, 0.0}, {1.0, 0.0}, {0.0, 1.0}, {-1.0, 0.0}};
    return ArrayGraph(std::move(offsets), std::move(targets),
                      std::move(weights), std::move(coords));
}

TEST_CASE("emit_geometry off leaves geometry empty", "[geometry]") {
    auto path = make_path5_with_coords();
    auto result = contract_degree2(path, {}, {}, /*emit_geometry=*/false);
    REQUIRE(result.edge_geometry.empty());
    REQUIRE(result.edge_geometry.edge_count() == 0);
}

TEST_CASE("degree-2 chain collapses to a real polyline", "[geometry]") {
    auto path = make_path5_with_coords();
    auto result = contract_degree2(path, {}, {}, /*emit_geometry=*/true);

    const auto& g = *result.graph;
    const auto& geom = result.edge_geometry;

    // Two surviving junctions (0, 4), two directed merged edges between them.
    REQUIRE(g.node_count() == 2);
    REQUIRE(g.edge_count() == 2);

    // Geometry is CSR-aligned to the edge set.
    REQUIRE(geom.edge_count() == g.edge_count());
    REQUIRE(geom.offsets.size() == g.edge_count() + 1);
    REQUIRE(geom.offsets.front() == 0);
    REQUIRE(geom.offsets.back() == geom.points.size());

    // The whole chain is preserved: 5 points per direction, 10 total.
    REQUIRE(geom.points.size() == 10);

    // Each edge's polyline starts at its source coord and ends at its target
    // coord, and both edges carry the full 5-point chain.
    auto off = g.raw_offsets();
    auto tg = g.raw_targets();
    for (NodeID u = 0; u < g.node_count(); ++u) {
        for (uint32_t e = off[u]; e < off[u + 1]; ++e) {
            NodeID v = tg[e];
            auto pts = geom.points_for(e);
            REQUIRE(pts.size() == 5);
            Coord src = g.node_coordinate(u).value();
            Coord dst = g.node_coordinate(v).value();
            REQUIRE(pts.front().lat == src.lat);
            REQUIRE(pts.front().lon == src.lon);
            REQUIRE(pts.back().lat == dst.lat);
            REQUIRE(pts.back().lon == dst.lon);
        }
    }

    // The two directed edges are reverses of each other.
    auto a = geom.points_for(0);
    auto b = geom.points_for(1);
    REQUIRE(a.size() == b.size());
    for (size_t i = 0; i < a.size(); ++i) {
        REQUIRE(a[i].lon == b[b.size() - 1 - i].lon);
    }
}

TEST_CASE("non-chain edges get 2-point geometry", "[geometry]") {
    auto star = make_star_with_coords();
    auto result = contract_degree2(star, {}, {}, /*emit_geometry=*/true);

    const auto& g = *result.graph;
    const auto& geom = result.edge_geometry;

    REQUIRE(g.node_count() == 4);          // nothing contracts
    REQUIRE(geom.edge_count() == g.edge_count());

    auto off = g.raw_offsets();
    auto tg = g.raw_targets();
    for (NodeID u = 0; u < g.node_count(); ++u) {
        for (uint32_t e = off[u]; e < off[u + 1]; ++e) {
            NodeID v = tg[e];
            auto pts = geom.points_for(e);
            REQUIRE(pts.size() == 2);       // straight segment
            REQUIRE(pts.front().lat == g.node_coordinate(u).value().lat);
            REQUIRE(pts.back().lon == g.node_coordinate(v).value().lon);
        }
    }
}

TEST_CASE("one-way degree-2 chain emits no phantom reverse edge", "[geometry][simplify]") {
    // Theta graph, junctions 0 & 1. The chain through mid 2 is ONE-WAY (1->2->0); the chains
    // through mids 3 and 4 are bidirectional. Weight is used as an existence proxy nowhere:
    // the one-way chain must contribute only its real direction, with no zero-weight phantom.
    std::vector<uint32_t> offsets = {0, 2, 5, 6, 8, 10};
    std::vector<NodeID> targets = {3, 4, /*1*/ 2, 3, 4, /*2*/ 0, /*3*/ 0, 1, /*4*/ 0, 1};
    std::vector<Weight> weights(targets.size(), 1.0);
    std::vector<Coord> coords = {{0, 0}, {2, 0}, {1, 0}, {1, 1}, {1, -1}};
    ArrayGraph g(std::move(offsets), std::move(targets), std::move(weights), std::move(coords));

    auto res = contract_degree2(g);  // emit_geometry defaults on
    const auto& s = *res.graph;

    REQUIRE(s.node_count() == 2);
    // 5 real directed edges: 0->1 (chains 3,4) and 1->0 (chains 2,3,4). The old
    // weight-inferred logic would add a 6th, a zero-weight phantom 0->1 for the one-way chain.
    REQUIRE(s.edge_count() == 5);
    for (Weight w : s.raw_weights()) REQUIRE(w > 0.0);  // no zero-weight phantom
    REQUIRE(res.edge_geometry.edge_count() == s.edge_count());  // geometry stays aligned
}

TEST_CASE("simplify_edge_geometry keeps off-chord points and drops near-chord ones",
          "[geometry][simplify]") {
    // Edge 0: bent 3-point polyline — midpoint (1,1) sits perpendicular distance 1.0 off the
    // chord (0,0)->(0,2). Edge 1: an already-straight 2-point segment (must be untouched).
    EdgeGeometry geom;
    geom.offsets = {0, 3, 5};
    geom.points = {{0.0, 0.0}, {1.0, 1.0}, {0.0, 2.0}, {5.0, 5.0}, {5.0, 7.0}};

    SECTION("tolerance below the off-chord distance retains the bend") {
        auto out = simplify_edge_geometry(geom, 0.5);  // 0.5 < 1.0 -> keep midpoint
        REQUIRE(out.edge_count() == 2);
        REQUIRE(out.points_for(0).size() == 3);  // bend retained
        REQUIRE(out.points_for(0)[1].lat == 1.0);
        REQUIRE(out.points_for(1).size() == 2);  // straight edge unchanged
    }

    SECTION("tolerance above the off-chord distance collapses to the endpoints") {
        auto out = simplify_edge_geometry(geom, 2.0);  // 2.0 > 1.0 -> drop midpoint
        REQUIRE(out.edge_count() == 2);
        REQUIRE(out.points_for(0).size() == 2);
        REQUIRE(out.points_for(0).front().lat == 0.0);
        REQUIRE(out.points_for(0).back().lon == 2.0);
        REQUIRE(out.points_for(1).size() == 2);
    }

    SECTION("tolerance <= 0 returns an unchanged copy") {
        auto out = simplify_edge_geometry(geom, 0.0);
        REQUIRE(out.points.size() == geom.points.size());
        REQUIRE(out.points_for(0).size() == 3);
    }
}

File: python/tests/test_edge_geometry.py
"""Tests for Phase 2B edge geometry: degree-2 contraction preserving each edge's
real polyline, and ``interop.to_geodataframe(..., edge_geometry=...)``.

The core simplify tests need only numpy; the GeoDataFrame tests are skipped
without geopandas/shapely; the Swain test is skipped without an OSM build."""

from pathlib import Path

import gravel
import numpy as np
import pytest

DATA_DIR = Path(__file__).resolve().parents[2] / "tests" / "data"
SWAIN_PBF = DATA_DIR / "swain_county.osm.pbf"


def _maybe_import(name):
    try:
        return __import__(name)
    except ImportError:
        return None


geopandas = _maybe_import("geopandas")
requires_geopandas = pytest.mark.skipif(geopandas is None, reason="geopandas not installed")
requires_osm = pytest.mark.skipif(
    not gravel.HAS_OSM or not SWAIN_PBF.exists(),
    reason="OSM support not built or Swain fixture absent (e.g. sdist without tests/data)",
)


def _theta_graph():
    """Two junctions (0, 1) joined by three parallel chains through degree-2
    mids (2, 3, 4). The chains form cycles, so no edge is a bridge and the mids
    are contractible even under the default bridge protection."""
    und = [(0, 2), (2, 1), (0, 3), (3, 1), (0, 4), (4, 1)]
    edges = und + [(b, a) for a, b in und]
    s = np.array([a for a, b in edges], np.uint32)
    t = np.array([b for a, b in edges], np.uint32)
    w = np.ones(len(edges))
    coords = np.array([[0, 0], [2, 0], [1, 1], [1, 0], [1, -1]], float)
    return gravel.Graph.from_coo(5, s, t, w, coords)


def _simplify(g, emit_geometry):
    cfg = gravel.SimplificationConfig()
    cfg.emit_geometry = emit_geometry
    cfg.estimate_degradation = False
    return gravel.simplify_graph(g, None, None, cfg)


def test_emit_geometry_off_is_empty():
    res = _simplify(_theta_graph(), emit_geometry=False)
    assert res.edge_geometry.empty
    assert res.edge_geometry.edge_count == 0


def test_degree2_contraction_preserves_polylines():
    res = _simplify(_theta_graph(), emit_geometry=True)
    geom = res.edge_geometry

    # Three chains collapse to two junctions with six directed merged edges.
    assert res.simplified_nodes == 2
    assert res.simplified_edges == 6
    assert not geom.empty
    assert geom.edge_count == res.simplified_edges
    assert geom.offsets[0] == 0
    assert geom.offsets[-1] == geom.points.shape[0]
    # Every merged edge keeps its 3-point chain (junction, mid, junction).
    assert [len(geom.polyline(e)) for e in range(geom.edge_count)] == [3] * 6


def test_polyline_endpoints_match_node_coords():
    res = _simplify(_theta_graph(), emit_geometry=True)
    geom = res.edge_geometry
    sources, targets, _ = res.graph.to_coo()
    node_xy = res.graph.node_coordinates()  # (N, 2) [lat, lon]
    for e in range(geom.edge_count):
        line = geom.polyline(e)  # (k, 2) [lat, lon]
        np.testing.assert_allclose(line[0], node_xy[sources[e]])
        np.testing.assert_allclose(line[-1], node_xy[targets[e]])


@requires_geopandas
def test_to_geodataframe_uses_real_geometry():
    res = _simplify(_theta_graph(), emit_geometry=True)
    gdf = gravel.interop.to_geodataframe(res.graph, edge_geometry=res.edge_geometry)
    assert len(gdf) == res.simplified_edges
    # Every drawn edge is a bent 3-vertex line, not a straight chord.
    assert all(len(line.coords) == 3 for line in gdf.geometry)
    # Coordinates are (x=lon, y=lat): mid node 4 is (lat=1, lon=-1) -> (-1, 1).
    verts = {tuple(c) for line in gdf.geometry for c in line.coords}
    assert (-1.0, 1.0) in verts


@requires_geopandas
def test_to_geodataframe_falls_back_to_straight():
    res = _simplify(_theta_graph(), emit_geometry=True)
    gdf = gravel.interop.to_geodataframe(res.graph)  # no edge_geometry
    assert all(len(line.coords) == 2 for line in gdf.geometry)


@requires_geopandas
def test_edge_geometry_length_mismatch_raises():
    g = _theta_graph()
    res = _simplify(g, emit_geometry=True)  # 6 edges of geometry
    # Applying it to the unsimplified graph (12 edges) must be rejected.
    with pytest.raises(ValueError):
        gravel.interop.to_geodataframe(g, edge_geometry=res.edge_geometry)


@requires_osm
@requires_geopandas
def test_swain_real_geometry():
    g, _ = gravel.datasets.osm.load_with_metadata(str(SWAIN_PBF))
    res = _simplify(g, emit_geometry=True)
    geom = res.edge_geometry
    assert geom.edge_count == res.simplified_edges
    # Real roads have curvature: at least some collapsed edges carry >2 points.
    assert any(len(geom.polyline(e)) > 2 for e in range(geom.edge_count))
    gdf = gravel.interop.to_geodataframe(res.graph, edge_geometry=geom)
    assert len(gdf) == res.simplified_edges
    assert gdf.geometry.is_valid.all()


def test_ch_pruning_clears_geometry():
    # CH-level pruning rebuilds the edge set; degree-2 geometry would misalign, so the
    # pipeline drops it. Assert: present without pruning, empty with pruning (never stale).
    g = _theta_graph()
    ch = gravel.build_ch(g)
    idx = gravel.ShortcutIndex(ch)
    cfg = gravel.SimplificationConfig()
    cfg.emit_geometry = True
    cfg.estimate_degradation = False
    assert not gravel.simplify_graph(g, None, None, cfg).edge_geometry.empty
    cfg.ch_level_keep_fraction = 0.5
    assert gravel.simplify_graph(g, ch, idx, cfg).edge_geometry.empty


def test_simplify_edge_geometry_reduces_points():
    # Theta graph -> 3-point bent merged edges; Douglas-Peucker with a large tolerance drops the
    # off-chord midpoints down to straight 2-point edges; tolerance 0 leaves it unchanged.
    from gravel import _gravel

    res = _simplify(_theta_graph(), emit_geometry=True)
    geom = res.edge_geometry
    before = geom.points.shape[0]

    collapsed = _gravel.simplify_edge_geometry(geom, 5.0)
    assert collapsed.edge_count == geom.edge_count
    assert collapsed.points.shape[0] < before
    assert all(len(collapsed.polyline(e)) == 2 for e in range(collapsed.edge_count))

    # Retention path: a tolerance below the off-chord midpoint distance (1.0 on the bent chains)
    # keeps those bends while still dropping the one collinear midpoint — so it lands strictly
    # between full collapse and the untouched geometry, and at least one polyline keeps its 3rd point.
    retained = _gravel.simplify_edge_geometry(geom, 0.5)
    assert collapsed.points.shape[0] < retained.points.shape[0] < before
    assert any(len(retained.polyline(e)) == 3 for e in range(retained.edge_count))

    unchanged = _gravel.simplify_edge_geometry(geom, 0.0)
    assert unchanged.points.shape[0] == before
